server/datastore/mysql/migrations/tables/20230315163954_SoftwareCPEUnqConstraint.go

```diff
@@ -0,0 +1,82 @@
+package tables
+
+import (
+	"database/sql"
+
+	"github.com/fleetdm/fleet/v4/server"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20230315163954, Down_20230315163954)
+}
+
+func Up_20230315163954(tx *sql.Tx) error {
+	// Since we will be adding a uniqueness constrain on (software_id) on the software_cpe table - we need to remove any
+	// possible duplicates. Also because there's a chance we remove the duplicate rows before adding
+	// the constraint and new duplicates get generated in between, we need to try to acquire the
+	// vulnerability lock. In case the lock can't be acquired a warning is issued and the migration
+	// will proceed without it.
+
+	removeDuplicatedSoftwareCPEs := func(tx *sql.Tx) error {
+		const deleteStmt = `
+DELETE sc
+FROM software_cpe sc
+	INNER JOIN (
+		SELECT
+			software_id,
+			MAX(id) as max_id
+		FROM software_cpe
+		GROUP BY software_id
+		HAVING COUNT(*) > 1
+	) sc2 ON sc2.software_id = sc.software_id
+WHERE sc.id < sc2.max_id;`
+
+		if _, err := tx.Exec(deleteStmt); err != nil {
+			return errors.Wrap(err, "removing duplicated rows")
+		}
+
+		return nil
+	}
+
+	addConstraint := func(tx *sql.Tx) error {
```

Comment:
Same here.
embedded functions are cool but I think they frequently reduce code readability.

Reply:
See previous comment, I can rename it to something else if you think that will help with readability. One thing to keep in mind is that this is a migration, this code will be run once and then never modified.

Reply:
I definitely think it's preferable. (Even for migration code)